Fix the double slash namespace bug (#2436)
treeverse/lakeFS: docs/reference/upgrade.md

#### Migrating Google Storage data with gsutil
[gsutil](https://cloud.google.com/storage/docs/gsutil) is a Python application that lets you access Cloud Storage from the command line.
We can use it for copying the data between the prefixes in the Google bucket, and later on removing it.

Copy the data with:
```shell
gsutil -m cp -r gs://<BUCKET>//<PREFIX>/ gs://<BUCKET>/
```

#### Migrating Azure Blob Storage data with AzCopy
[AzCopy](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-v10) is a command-line utility that you can use to copy blobs or files to or from a storage account.
We can use it for copying the data between the prefixes in the Azure storage account container, and later on removing it.

First, you need to acquire an [Account SAS](https://docs.microsoft.com/en-us/azure/storage/common/storage-sas-overview#account-sas).
Using the Azure CLI:
```shell
az storage container generate-sas \
    --account-name <ACCOUNT> \
    --name <CONTAINER> \
    --permissions cdrw \
    --auth-mode key \

>     --expiry 2021-09-05

Reviewer: Nonetheless, this expiry has elapsed.  Options:
1. Point it out and ask people to fill in a good date.
1. Set a date a month into the future.
1. Use GNU date (so bad news on Windows) and suggest ``--expiry `date -I --date '+1 day'` `` (_should_ also work on MacOS date, it tries to be GNUish, but check that first...).

cataloger:
  type: rocks
```

> ## Data Migration for Version v0.49.0

Reviewer: Reminder: as discussed F2F I had to use up this minor version on Thursday.  You'll have to bump in

We discovered a bug in the way lakeFS is storing objects in the underlying object store.

> It affects azure and gcp repositories and only some of those. 

Reviewer: ```suggestion
It affects  only repositories on Azure and GCP, and not all of these.
```

> It affects only repositories on Azure and GCP, and not all of these.

Reviewer 2: ```suggestion
It affects only some repositories on Azure and GCP. Repositories in S3 were not affected.
```

[Issue #2397](https://github.com/treeverse/lakeFS/issues/2397#issuecomment-908397229) describes the repository storage namespaces patterns 
which are affected by this bug.

> In order to upgrade to any version which higher (or equal) than v0.50.0, you must follow these steps:

Reviewer 2: ```suggestion
When first updating to a version greater or equal to v0.50.0, you must follow these steps:
```

1. Stop lakeFS.
1. Perform a data-migration (details below)
1. Start lakeFS with the new version.
1. After a successful run of the new version, and after validating the objects are accessible, you can delete the old data prefix.

### Data migration

> The following patterns have been impacted by the bug:

Reviewer 2: ```suggestion
To determine if your repositories were affected by this bug, take a look at their _storage namespace_. You can find it in the repository settings page in the UI, or using `lakectl repo list`.

The following storage namespace patterns have been impacted by the bug:
```

| gs    | gs://bucket/prefix/                                       | gs://bucket//prefix/*                                      | gs://bucket/prefix/*                                       |
| azure | https://account.blob.core.windows.net/containerid         | https://account.blob.core.windows.net/containerid//*       | https://account.blob.core.windows.net/containerid/*        |
| azure | https://account.blob.core.windows.net/containerid/        | https://account.blob.core.windows.net/containerid//*       | https://account.blob.core.windows.net/containerid/*        |
| azure | https://account.blob.core.windows.net/containerid/prefix/ | https://account.blob.core.windows.net/containerid/prefix// | https://account.blob.core.windows.net/containerid/prefix/* |

#### Migrating Google Storage data with gsutil

[gsutil](https://cloud.google.com/storage/docs/gsutil) is a Python application that lets you access Cloud Storage from the command line.
We can use it for copying the data between the prefixes in the Google bucket, and later on removing it.

> Copy the data with:

Reviewer 2: ```suggestion
For every affected repository, copy its data with:
```

```shell
gsutil -m cp -r gs://<BUCKET>//<PREFIX>/ gs://<BUCKET>/
```

> 

Reviewer 2: ```suggestion
Note the double slash after the bucket name.
```

[gsutil](https://cloud.google.com/storage/docs/gsutil) is a Python application that lets you access Cloud Storage from the command line.
We can use it for copying the data between the prefixes in the Google bucket, and later on removing it.

For every affected repository, copy its data with:
```shell
gsutil -m cp -r gs://<BUCKET>//<PREFIX>/ gs://<BUCKET>/
```

Note the double slash after the bucket name.

#### Migrating Azure Blob Storage data with AzCopy

[AzCopy](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-v10) is a command-line utility that you can use to copy blobs or files to or from a storage account.
We can use it for copying the data between the prefixes in the Azure storage account container, and later on removing it.

First, you need to acquire an [Account SAS](https://docs.microsoft.com/en-us/azure/storage/common/storage-sas-overview#account-sas).
Using the Azure CLI:
```shell
az storage container generate-sas \
    --account-name <ACCOUNT> \
    --name <CONTAINER> \
    --permissions cdrw \
    --auth-mode key \
    --expiry 2021-12-31
```

With the resulted SAS, use AzCopy to copy the files:
```shell

> azcopy copy \

Reviewer 2: What is the command for when there is no prefix?
